device_attributes/index.bs

```diff
@@ -0,0 +1,192 @@
+ 
+### <dfn>Hostname</dfn> ### {#hostname}
+ 
+An administrator-defined value which is used as the device hostname during DHCP requests.
+ 
+### <dfn>Serial Number</dfn> ### {#serial-number}
+ 
+A manufacturer-defined value which uniquely identifies a device among those produced by that manufacturer.
+ 
+# {{NavigatorManagedData}} interface # {#navigatormanageddata-interface}
+ 
+<xmp class="idl">
+[
+  SecureContext,
+  Exposed=Window
+] interface NavigatorManagedData : EventTarget {
+  // Device Attributes API.
+  Promise<DOMString> getAnnotatedAssetId();
+  Promise<DOMString> getAnnotatedLocation();
+  Promise<DOMString> getDirectoryId();
+  Promise<DOMString> getHostname();
+  Promise<DOMString> getSerialNumber();
+};
+</xmp>
+ 
+Methods on this interface typically complete asynchronously, queuing work on the <dfn>managed data task source</dfn>.
+ 
+## {{getAnnotatedAssetId()}} method ## {#getAnnotatedAssetId-method}
+ 
+  The <dfn method for="NavigatorManagedData">getAnnotatedAssetId()</dfn> method steps are:
+  1. Let |promise| be [=a new promise=].
+  1. Run the following steps in parallel:
+      1. If this API is not called by a managed web application with permission to access device attributes, reject |promise| with a {{NotAllowedError}} DOMException.
+      1. If no [=Annotated Asset ID=] has been set by the administrator, resolve |promise| with {{undefined}} value.
+      1. [=Queue a global task=] on the [=relevant global object=] of [=this=] using the [=managed data task source=] to [=resolve=] promise with the [=Annotated Asset ID=].
+  1. Return |promise|.
+ 
+## {{getAnnotatedLocation()}} method ## {#getAnnotatedLocation-method}
+ 
+  The <dfn method for="NavigatorManagedData">getAnnotatedLocation()</dfn> method steps are:
+  1. Let |promise| be [=a new promise=].
+  1. Run the following steps in parallel:
+      1. If this API is not called by a managed web application with permission to access device attributes, reject |promise| with a {{NotAllowedError}} DOMException.
+      1. If no [=Annotated Location=] has been set by the administrator, resolve |promise| with {{undefined}} value.
+      1. [=Queue a global task=] on the [=relevant global object=] of [=this=] using the [=managed data task source=] to [=resolve=] promise with the [=Annotated Location=].
+  1. Return |promise|.
+ 
+## {{getDirectoryId()}} method ## {#getDirectoryId-method}
+ 
+  The <dfn method for="NavigatorManagedData">getDirectoryId()</dfn> method steps are:
+  1. Let |promise| be [=a new promise=].
+  1. Run the following steps in parallel:
+      1. If this API is not called by a managed web application with permission to access device attributes, reject |promise| with a {{NotAllowedError}} DOMException.
+      1. [=Queue a global task=] on the [=relevant global object=] of [=this=] using the [=managed data task source=] to [=resolve=] promise with the [=Directory ID=].
```

Comment:
This assumes that every device has a directory ID, which probably isn't true.

Reply:
The assumption, which may be valid, is that any managed device will have a directory ID and to get to this point in the algorithm the device must be managed.

I raised this in one of my earlier issues and it came up again in #11 that these properties make assumptions about the management infrastructure which could be avoided if we didn't try to standardize particular device properties and instead relied only on [[MANAGED-CONFIG]](https://wicg.github.io/WebApiDevice/managed_config/) to provide a communication channel between the management infrastructure and the managed application.

Reply:
We could also provide a registry of some sort for common [MANAGED-CONFIG] keys, similar to the registries for [`<meta>` names](https://html.spec.whatwg.org/multipage/semantics.html#standard-metadata-names) and [`<link>` relations](https://html.spec.whatwg.org/multipage/links.html#linkTypes).

The management consoles, or even particular browsers, could inject values for some of those keys from other calls down to native APIs, like for the serial number and directory ID.

Reply:
@Ananubis, what do you think of this proposal?

Reply:
We considered this option at the beginning, but didn't pick it because the following reasons.
- It doesn't help to simplify the specification of various device attributes. Each reserved key still needs to be defined as what we are doing in the draft spacification now.
- If we integrate these device attributes APIs into the managed config API. When we want to add some new reserved key in the future, it is difficult to ensure the new key hasn't been used by the existing managed configuration. Such conflicts may break the existing web applications suddenly.
- In concept, **any getter web API** can be integrated into the managed config API if we want, not matter it is a device name, a display name, the current cursor's location or something else. But an API call like getManagedConfiguration("serialNumber") is not intuitive as getSerialNumber(), since each device usually has a serial number.
- The managed config API doesn't require managed (force-installed) application, but the device attributes API does.